state-synchronizer/src/coordinator.rs

```diff
@@ -361,163 +426,191 @@ impl<T: ExecutorProxyTrait> SyncCoordinator<T> {
-            self.peer_manager.process_response(version, *peer_id);
-
-            if version != self.known_version + 1 {
-                // version was not requested, or version was requested from a different peer,
-                // so need to penalize peer for maliciously sending chunk
-                if has_requested {
-                    self.peer_manager
-                        .update_score(&peer_id, PeerScoreUpdateType::InvalidChunk)
-                }
-                return Err(format_err!(
-                    "[state sync] non sequential chunk. Known version: {}, received: {}",
-                    self.known_version,
-                    version,
-                ));
-            }
+        let known_version = self.highest_version_in_local_storage();
+        if txn_list_with_proof.is_empty() {
+            self.peer_manager
+                .update_score(peer_id, PeerScoreUpdateType::EmptyChunk);
+            self.send_chunk_request(known_version).await; // retry without waiting for timeout
+            bail!("[state sync] Empty chunk from {}", peer_id.short_str());
         }
 
-        let previous_version = self.known_version;
-        let chunk_size = txn_list_with_proof.len();
-        let target: LedgerInfo = response
+        let chunk_start_version = txn_list_with_proof.first_transaction_version.unwrap_or(0);
+        if !self
+            .peer_manager
+            .has_requested(chunk_start_version, *peer_id)
+        {
+            self.peer_manager
+                .update_score(peer_id, PeerScoreUpdateType::InvalidChunk);
+            bail!(
+                "[state sync] Chunk with version {} was not requested from {}",
+                chunk_start_version,
+                peer_id.short_str()
+            );
+        }
+        self.peer_manager
+            .process_response(chunk_start_version, *peer_id);
+        if chunk_start_version != known_version + 1 {
+            bail!(
+                "[state sync] Non sequential chunk from {}: known_version: {}, received: {}",
+                peer_id.short_str(),
+                known_version,
+                chunk_start_version
+            );
+        }
+        let response_li: LedgerInfoWithSignatures = response
             .ledger_info_with_sigs
             .ok_or_else(|| format_err!("Missing ledger_info_with_sigs"))?
             .try_into()?;
 
-        let result = self
-            .validate_and_store_chunk(txn_list_with_proof, target.clone())
-            .await;
-        let latest_version = self.executor_proxy.get_latest_version().await?;
-        if latest_version <= previous_version {
+        if let Some(sync_req) = self.sync_request.as_ref() {
+            let response_li_version = response_li.ledger_info().version();
+            // Valid responses must come with the LI version of the request.
+            if sync_req.target.ledger_info().version() != response_li_version {
+                self.peer_manager
```

Comment:
not for this PR, but we probably can discuss how to improve the peer_manager's API, currently it's entangled with the code and not look very nice